src/secret/mod.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Secret manager module enabling address generation and transaction essence signing.
+
+#[cfg(feature = "ledger")]
+pub mod ledger;
+/// Module for signing with a mnemonic or seed
+pub mod mnemonic;
+/// Module for signing with a Stronghold vault
+#[cfg(feature = "stronghold")]
+pub mod stronghold;
+/// Signing related types
+pub mod types;
+
+#[cfg(feature = "wasm")]
+use std::sync::Mutex;
+use std::{collections::HashMap, ops::Range, path::PathBuf, str::FromStr};
+
+use async_trait::async_trait;
+use bee_message::{
+    address::{Address, AliasAddress, Ed25519Address, NftAddress},
+    output::Output,
+    payload::transaction::TransactionEssence,
+    unlock_block::{AliasUnlockBlock, NftUnlockBlock, ReferenceUnlockBlock, UnlockBlock},
+};
+pub use types::{GenerateAddressMetadata, LedgerStatus, Network, SignMessageMetadata};
+
+#[cfg(feature = "ledger")]
+use self::ledger::LedgerSecretManager;
+#[cfg(feature = "stronghold")]
+use self::stronghold::StrongholdSecretManager;
+use self::{mnemonic::MnemonicSecretManager, types::SecretManagerDto};
+use crate::secret::types::InputSigningData;
+
+/// The secret manager interface.
+#[async_trait]
+pub trait SecretManage: Send + Sync {
+    /// Generates an address.
+    ///
+    /// For `coin_type`, see also <https://github.com/satoshilabs/slips/blob/master/slip-0044.md>.
+    async fn generate_addresses(
+        &self,
+        coin_type: u32,
+        account_index: u32,
+        address_indexes: Range<u32>,
+        internal: bool,
+        metadata: GenerateAddressMetadata,
+    ) -> crate::Result<Vec<Address>>;
+
+    /// Sign on `essence`, unlock `input` by returning an [UnlockBlock].
+    ///
+    /// Compared with `sign_transaction_essence()`, this signs on a single input. It's called by
+    /// `sign_transaction_essence()` on every input it receives.
```

Comment:
> on every input it receives.

Remove this part or rewrite, because it's only called if the input requires an Ed25519 signature and if there is no signature for the address already